src/dali_executor/io_buffer.h

```diff
@@ -0,0 +1,155 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef TRITONDALIBACKEND_IO_BUFFER_H
+#define TRITONDALIBACKEND_IO_BUFFER_H
+
+#include "src/dali_executor/io_descriptor.h"
+#include "src/dali_executor/utils/dali.h"
+
+namespace triton { namespace backend { namespace dali {
+
+
+void MemCopy(device_type_t dst_dev, void *dst, device_type_t src_dev, const void *src, size_t size,
+             cudaStream_t stream = 0);
+
+class IOBufferI {
+ public:
+  /**
+   * @brief Allocate a chunk of `size` of bytes and return a pointer
+   *        to the beginning of the chunk.
+   * @param size Size of the allocation in bytes.
+   * @return Pointer to the beggining of the allocated chunk.
+   */
+  virtual uint8_t *Allocate(size_t size) = 0;
+
+  /**
+   * @brief Cancel all reservations. No memory is deallocated.
+   */
+  virtual void Clear() = 0;
+
+  /**
+   * @brief Reserve `size` bytes of memory.
+   *
+   * If the buffer's capacity is greater or equal to size, this function is a no-op.
+   * @param size Amount of memory to reserve in bytes.
+   */
+  virtual void Reserve(size_t size) = 0;
+
+  /**
+   * @brief Get the the amount of allocated memory.
+   * @return Allocation size in bytes.
+   */
+  virtual size_t Capacity() const = 0;
+
+  /**
+   * @brief Get device type of the allocated memory.
+   * @return Device type.
+   */
+  virtual device_type_t DeviceType() const = 0;
+
+  /**
+   * @brief Get an immutable descriptor of the buffer.
+   * @return Input buffer descriptor.
+   */
+  virtual IBufferDescr GetDescr() const = 0;
+
+  /**
+   * @brief Get a mutable descriptor of the buffer.
+   * @return Output buffer descriptor.
+   */
+  virtual OBufferDescr GetDescr() = 0;
+
+  virtual ~IOBufferI() {}
+
+ protected:
+  IOBufferI() {}
+};
+
+template<typename T, device_type_t Dev>
+using buffer_t = std::conditional_t<Dev == device_type_t::CPU, std::vector<T>, DeviceBuffer<T>>;
+
+template<device_type_t Dev>
+class IOBuffer : public IOBufferI {
+ public:
+  IOBuffer(size_t size = 0) : buffer_() {
+    buffer_.resize(size);
+    if (Dev == device_type_t::GPU) {
+      CUDA_CALL_GUARD(cudaGetDevice(&device_id_));
+    }
+  }
+
+  uint8_t *Allocate(size_t size) override {
```

Comment:
1. The name of this function is unfortunate to say the least. Without reading the code `Allocate` and `Reserve` seem like synonyms.
2. There's no alignment support. x86 or ARM don't require aligned memory to work _at all_, but they do require natural alignment to perform well. I think that the GPU requires natural alignment at least for fundamental types, though. Most likely this worked by sheer luck.

```suggestion
  uint8_t *GetChunk(size_t size, size_t alignment = alignof(std::max_align_t)) override {
```